lib/get-eslint-modules.js

```diff
@@ -0,0 +1,44 @@
+var Module = require("module");
+var path = require("path");
+
+/**
+ * We can't just require in libraries related to ESLint, scope traversal etc.
+ * We need to look them up based on what the end user has installed.
+ *
+ * This function handles tracking down all the relevant libraries in this area
+ * and returns an object which includes them as properties.
+ */
+module.exports = function getESLintModules() {
+  try {
+    // avoid importing a local copy of eslint, try to find a peer dependency
+    var eslintLoc = Module._resolveFilename("eslint", module.parent);
+  } catch (err) {
+    try {
+      // avoids breaking in jest where module.parent is undefined
+      eslintLoc = require.resolve("eslint");
+    } catch (err) {
+      throw new ReferenceError("couldn't resolve eslint");
+    }
+  }
+
+  // get modules relative to what eslint will load
+  var eslintMod = new Module(eslintLoc);
+  eslintMod.filename = eslintLoc;
+  eslintMod.paths = Module._nodeModulePaths(path.dirname(eslintLoc));
+
+  var eslintScope = eslintMod.require("eslint-scope");
+  var Definition = eslintMod.require("eslint-scope/lib/definition").Definition;
+  var Referencer = eslintMod.require("eslint-scope/lib/referencer");
+  var Traverser = eslintMod.require("eslint/lib/util/traverser");
+  var ScopeManager = eslintMod.require("eslint-scope/lib/scope-manager");
```

Comment:
I really like and agreed with that perspective, @not-an-aardvark!

I don't think depending on eslint-scope or estraverse directly is an option, because they need to be the matching versions to the end user's installed ESLint version. We would be leaving ourselves open to issues there if eslint-scope was a dep, but ESLint was a resolved peerDep.

I would like to hear from @mysticatea on this too, of course, because I believe I have implemented this how he originally intended in his ESLint PR, but I may be wrong.

I believe we would want to extend the ESLint PR further to provide APIs to solve the fragility you mention, and as I say I think you've made an excellent point.